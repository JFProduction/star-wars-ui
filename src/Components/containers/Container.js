import React, { Component } from 'react'
import Loading from '../presentations/Loading'
import PageNav from '../presentations/PageNav'
import ViewType from '../presentations/ViewType'
import ListHeader from '../presentations/ListHeader'
import SearchStuff from '../presentations/SearchStuff'

import { bindActionCreators } from "redux"
import { connect } from "react-redux"
import { 
  requestApiData, 
  removePerson
} from "../../sagas/actions";
import Layout from '../presentations/Layout';
import { ModalPres } from '../presentations/ModalPres'

export class Container extends Component {
  constructor(props) {
    super(props)
    this.state = {
      layout: "card"
    }

    this.inputRef = React.createRef()
  }

  componentDidMount() {
    this.props.requestApi("https://swapi.co/api/people")
  }

  clickNextPrev = which => _ => {
    const url = which === 1 ? this.props.people.next : this.props.people.previous
    this.props.requestApi(url)
  }

  handleClose = () => {
    this.props.removePerson()
  }

  handleClickViewType = ({target}) => {
    this.setState({layout: target.id})
  }
  
  render() {
    const {
      layout
    } = this.state

    const { people, selectedPerson } = this.props

    return (
      <div
        style={{position: "relative"}}
      >
        <SearchStuff
          search={this.search}
          inputRef={this.inputRef}
        />
        {
          people &&
            <PageNav
              onClick={this.clickNextPrev}
              prev={people.previous}
              next={people.next}
              page={0}
              loading={false}
            />
        }
        <ViewType
          handleClick={this.handleClickViewType}
          view={layout}
          value={263}
        />
      {
        selectedPerson.films && 
          <ModalPres
            open={true}
            person={selectedPerson}
            handleClose={this.handleClose}
          />
      }
      {
        layout === "list" && (
          <ListHeader
            headerList={[
              {lbl: "name", style: {marginLeft: 255}},
              {lbl: "gender", style: {marginLeft: 117}},
              {lbl: "height", style: {marginLeft: 120}},
              {lbl: "mass", style: {marginLeft: 118}},
            ]}
            headerClick={this.handleClose}
            headerStyle={{marginBottom: 12}}
          />
        )
      }
      {
        people.results && people.results.length > 0 ?
          people.results.map((person, i) => {
            return (
              person.name && 
                <Layout
                  person={person}
                  key={i}
                  layout={layout}
                />
            )
          })
          : <Loading />
      }
      </div>
    )
  }
}

const mapStateToProps = ({people, selectedPerson}) => {
  return {people, selectedPerson}
}

const mapDispatchToProps = dispatch => bindActionCreators({ 
  requestApi: url => requestApiData(url),
  removePerson,
}, dispatch)

export default connect(
  mapStateToProps, 
  mapDispatchToProps
)(Container)