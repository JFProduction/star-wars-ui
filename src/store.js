import { createStore, applyMiddleware } from "redux"
import createSagaMiddleware from "redux-saga"

import {DataReducer} from "./reducers/data"
import mySaga from "./sagas"


// create the saga middleware
const sagaMiddleware = createSagaMiddleware()
// mount it on the Store
export default createStore(
  DataReducer,
  applyMiddleware(sagaMiddleware)
)

// then run the saga
sagaMiddleware.run(mySaga)