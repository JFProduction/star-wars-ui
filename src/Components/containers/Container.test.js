import React from 'react'
import { mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import Container from './Container'
import { Provider } from 'react-redux'
import store from '../../store'
import resp from '../../../__mocks__/MockResult'

describe("container tests", () => {
  let wrapper

  beforeAll(() => {
    window.fetch = jest.fn(() => Promise.resolve({
      json: () => Promise.resolve(resp)
    }))

    wrapper = mount(
      <Provider store={store}>
        <Container />
      </Provider>
    )
  })

  it("should match snapshot", done => {
    setTimeout(() => {
      wrapper.update()
      expect(shallowToJson(wrapper)).toMatchSnapshot()
      done()
    }, 0)
  })

  it("should render 10 main cards", () => {
    expect(wrapper.find("MainCard")).toHaveLength(10)
  })

  // TODO: Figure this out...
  // it("should change state.layout to list", () => {
  //   expect(wrapper.state().layout).toEqual("card")
  //   wrapper.find("img[id='list']").simulate("click")
  //   wrapper.update()
  //   expect(wrapper.state().layout).toEqual("list")
  // })
})