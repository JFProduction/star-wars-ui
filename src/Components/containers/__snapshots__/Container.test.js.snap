// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container tests render tests should match snapshot 1`] = `
<Container>
  <div
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <form
      onSubmit={[Function]}
      style={
        Object {
          "position": "absolute",
          "right": 10,
          "top": -60,
        }
      }
    >
      <input
        placeholder="Global Search..."
        style={
          Object {
            "border": "1px solid #ddd",
            "borderRadius": 5,
            "marginRight": 10,
            "padding": 10,
            "width": 150,
          }
        }
      />
      <button
        className="btn"
        type="submit"
      >
        go
      </button>
    </form>
    <PageNav
      loading={true}
      next=""
      onClick={[Function]}
      page={0}
      prev=""
    >
      <div
        style={
          Object {
            "display": "flex",
            "margin": 10,
            "textAlign": "center",
          }
        }
      >
        <div
          style={
            Object {
              "width": "100%",
            }
          }
        />
        <div
          style={
            Object {
              "width": "100%",
            }
          }
        />
        <div
          style={
            Object {
              "width": "100%",
            }
          }
        />
      </div>
    </PageNav>
    <ViewType
      handleClick={[Function]}
      value={263}
      view="card"
    >
      <div
        style={
          Object {
            "background": "transparent",
            "position": "relative",
            "textAlign": "right",
            "width": "100%",
          }
        }
      >
        <img
          alt="Card View"
          className="view-btn active"
          id="card"
          onClick={[Function]}
          src="Icon-Cardview-Active.svg"
          title="Card View"
        />
        <img
          alt="List View"
          className="view-btn "
          id="list"
          onClick={[Function]}
          src="Icon-Listview.svg"
          title="List View"
        />
      </div>
    </ViewType>
    <Loading>
      <div
        className="wrapper"
      >
        <div
          className="one"
        />
        <div
          className="two"
        />
        <div
          className="three"
        />
        <div
          className="four"
        />
        <div
          className="lbl"
        >
          Loading...
        </div>
      </div>
    </Loading>
  </div>
</Container>
`;
